lib/RestructureCFG/DAGifyPass.cpp

```diff
@@ -0,0 +1,181 @@
+//
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/IR/IRBuilder.h"
+
+#include "revng/RestructureCFG/DAGifyPass.h"
+#include "revng/RestructureCFG/GenericRegionInfo.h"
+#include "revng/RestructureCFG/ScopeGraphGraphTraits.h"
+#include "revng/RestructureCFG/ScopeGraphUtils.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/GraphAlgorithms.h"
+#include "revng/Support/IRHelpers.h"
+
+using namespace llvm;
+
+// Debug logger
+Logger<> DAGifyPassLogger("dagify");
+
+/// Helper function which transform a retreating edge into a `goto` edge
+static void
+processRetreating(const revng::detail::FullEdgeDescriptor<BasicBlock *>
+                    &Retreating) {
```

Comment:
Se we're going to have a regular `EdgeDescriptor` here and we'll have to manually handle duplicated edges.

Reply:
Since we do not take into account the multiplicity of the edges anymore, as agreed, the algorithm now simply insert a `goto` block, and redirects all the multiple retreating edges there https://github.com/revng/revng/blob/4061da2db93fbe56637be5fe70488e17b66cbda1/lib/RestructureCFG/DAGifyPass.cpp#L63

In addition, I also took care of asserting that the retreating edges we are moving are not `scope_closer` edges, adding a disclaimer saying that this is not impossible in principle, just not expected now https://github.com/revng/revng/blob/4061da2db93fbe56637be5fe70488e17b66cbda1/lib/RestructureCFG/DAGifyPass.cpp#L38